doc/proposal/thrift-parameter-validation-proposal.md

```diff
@@ -0,0 +1,180 @@
+| ne           | not equals to (`!=`)             | i8, i16, i32, i64, f64, string, bool | same with value        | -                  |
+| lt           | less than (`<`)                  | i8, i16, i32, i64, f64               | same with value        | -                  |
+| le           | less equal (`<=`)                | i8, i16, i32, i64, f64               | same with value        | -                  |
+| gt           | greater than (`>`)               | i8, i16, i32, i64, f64               | same with value        | -                  |
+| ge           | greater equal (`>=`)             | i8, i16, i32, i64, f64               | same with value        | -                  |
+| in           | within given container           | i8, i16, i32, i64, f64               | same with value        | -                  |
+| not_in       | not within given container       | i8, i16, i32, i64, f64               | same with value        | -                  |
+| elem         | field's element constraint       | any                                  | list, set              | support            |
+| key          | field's element key constraint   | any                                  | map                    | support            |
+| value        | field's element value constraint | any                                  | map                    | support            |
+| min_size     | minimal length                   | i8, i16, i32, i64                    | string, list, set, map | -                  |
+| max_size     | maximal length                   | i8, i16, i32, i64                    | string, list, set, map | -                  |
+| prefix       | field prefix must be             | string                               | string                 | -                  |
+| suffix       | suffix must be                   | string                               | string                 | -                  |
+| contain      | must contain                     | string                               | string                 | -                  |
+| not_contain  | must not contain                 | string                               | string                 | -                  |
+
+- Secodary validator is a successive validator, usually used at container-type field. See below Set/List/Map examples.
+
+#### 3.2.2 IDL example
+- Number
+```
+struct NumericDemo{
+    1: double Value (validator.gte = "1000.1", validator.lte = "10000.1")
+    2: i8 Type (validator.in = "[1, 2, 4]")
+}
+```
+- String/Binary
+``` 
+struct StringDemo{
+    1: string Uninitialized (vt.const = "abc")
+    2: string Name (vt.min_size = "6", vt.max_size = "12")
+    3: string SomeStuffs (vt.pattern = "[0-9A-Za-z]+")
+    4: string DebugInfo (vt.prefix = "[Debug]")
+    5: string PanicInfo (vt.contains = "panic")
+}
+```
+- Bool
+```
+struct BoolDemo {
+    1: bool AMD (vt.const = "true")
+}
+```
+- Enum
+```
+struct EnumDemo {
+    1: Type AddressType (vt.const = "String")
```

Comment:
I guess here in this example `String` is one of the values of enum `Type`? it helps to add that into the example (add the definition of `Type` here).

Reply:
also according to the table `const` can only be used on string and bool types, while an enum is technically an int type. I also don't see why `in` and `not_in` can only be used for int types but not string/bool types. all in all I still don't see anything `const` can do that `in` cannot so I'd suggest just use `in` and remove `const`.

Reply:
`vt.int` and `vt.not` will be implemented in the form of an array, and the value of annotation is a string, which means that we need to escape special characters. In `vt.in`, we use `,` as separators which means we must use `"` or `'` to identify the value of the string, which means that we have to escape again, which will make `vt.in` and `vt.not_in` are very complex. That's why in previous versions we defined different values by defining multiple `vt.in` instead of defining multiple values in one `vt.in`.
To sum up, we think it is inappropriate to use `vt.in` or `vt.not_in` in string type. If there is such a requirement, it can be implemented by `pattern`.